File: src/main/java/br/unifor/costify/infra/data/entities/RecipeTable.java
package br.unifor.costify.infra.data.entities;

import br.unifor.costify.domain.entity.Recipe;
import br.unifor.costify.domain.valueobject.Id;
import br.unifor.costify.domain.valueobject.Money;
import br.unifor.costify.domain.valueobject.RecipeIngredient;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import org.hibernate.annotations.UpdateTimestamp;
import org.springframework.data.annotation.CreatedDate;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "recipes")
public class RecipeTable {
  @jakarta.persistence.Id
  public String id;

  @Column(nullable = false)
  public String name;

  @Column(nullable = false)
  public BigDecimal totalCost;

  @OneToMany(mappedBy = "recipeId", cascade = CascadeType.ALL, fetch = FetchType.EAGER, orphanRemoval = true)
  public List<RecipeIngredientTable> ingredients = new ArrayList<>();

  @CreatedDate
  public LocalDateTime createdAt;

  @UpdateTimestamp
  public LocalDateTime updatedAt;

  public static RecipeTable fromDomain(Recipe recipe) {
    RecipeTable table = new RecipeTable();
    table.id = recipe.getId().getValue();
    table.name = recipe.getName();
    table.totalCost = recipe.getTotalCost().getAmount();
    
    table.ingredients = recipe.getIngredients().stream()
        .map(ingredient -> RecipeIngredientTable.fromDomain(recipe.getId().getValue(), ingredient))
        .toList();
    
    return table;
  }

  public static Recipe toDomain(RecipeTable raw) {
    List<RecipeIngredient> ingredients = raw.ingredients.stream()
        .map(RecipeIngredientTable::toDomain)
        .toList();

    return new Recipe(
        Id.of(raw.id),
        raw.name,
        ingredients,
        Money.of(raw.totalCost)
    );
  }
}

File: internal/infrastructure/database/models/recipe_table.go
package models

import (
	"time"

	"github.com/vini/costify-go/internal/domain/entity"
	"github.com/vini/costify-go/internal/domain/valueobject"
)

// RecipeTable represents the recipes table model for GORM
type RecipeTable struct {
	ID        string    `gorm:"primaryKey;column:id"`
	Name      string    `gorm:"not null;unique;column:name"`
	TotalCost float64   `gorm:"not null;default:0;column:total_cost;check:total_cost >= 0"`
	CreatedAt time.Time `gorm:"default:CURRENT_TIMESTAMP;column:created_at"`
	UpdatedAt time.Time `gorm:"default:CURRENT_TIMESTAMP;column:updated_at"`

	// Relationships
	RecipeIngredients []RecipeIngredientTable `gorm:"foreignKey:RecipeID;constraint:OnDelete:CASCADE"`
}

// TableName returns the table name for GORM
func (RecipeTable) TableName() string {
	return "recipes"
}

// ToEntity converts GORM model to domain entity
func (rt *RecipeTable) ToEntity() (*entity.Recipe, error) {
	id := valueobject.Id{}.Of(rt.ID)

	var ingredients []valueobject.RecipeIngredient
	for _, ri := range rt.RecipeIngredients {
		ingredient, err := ri.ToValueObject()
		if err != nil {
			return nil, err
		}
		ingredients = append(ingredients, ingredient)
	}

	totalCost, err := valueobject.Money{}.Of(rt.TotalCost)
	if err != nil {
		return nil, err
	}

	return entity.NewRecipe(id, rt.Name, ingredients, totalCost)
}

// FromEntity converts domain entity to GORM model
func FromRecipeEntity(recipe *entity.Recipe) *RecipeTable {
	recipeTable := &RecipeTable{
		ID:        recipe.Id().Value(),
		Name:      recipe.Name(),
		TotalCost: recipe.TotalCost().Amount(),
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	for _, ingredient := range recipe.Ingredients() {
		recipeIngredient := FromRecipeIngredientValueObject(recipe.Id().Value(), ingredient)
		recipeTable.RecipeIngredients = append(recipeTable.RecipeIngredients, *recipeIngredient)
	}

	return recipeTable
}
